orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/core/graph/pipeline_transformer.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace training {
+
+void GetPipelineSendOutput(const Graph& graph, std::string& loss_name) {
+  for (auto& node : graph.Nodes()) {
+    if (!node.OpType().compare("Send")) {
+      // send op should always have an output, which is the OutputSignal.
+      loss_name = node.OutputDefs()[0]->Name();
+      return;
+    }
+  }
+}
+
+Status TransformGraphForPipeline(Graph& graph) {
+  // insert WaitEvent and RecordEvent to the partition
+  Node* send_fw{nullptr};
+  Node* send_bw{nullptr};
+  Node* recv_fw{nullptr};
+  Node* recv_bw{nullptr};
+  for (auto& node : graph.Nodes()) {
+    if (node.OpType() == "Send") {
+      if (node.Description() == "Backward pass") {
```

Comment:
>node.Description() == "Backward pass" [](start = 10, length = 37)

Can we make it a util function to check if node is in bw graph? We might change its representation later, for example, turning the string into json to carry more info.

Reply:
Good point. I'll change it.

---
In reply to: [408339164](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r408339164) [](ancestors = 408339164)